src/docs/autoscaling.md

```diff
@@ -0,0 +1,113 @@
+
+```
+desired_instance_count = max(request_counts)/request_per_instance
+```
+
+`request_per_instance` is a pre-defined scaler that indicates the number of requests that can be handled by one instance of app. 
+
+
+### Scheduled-jobs ##
+Notify uses a proactive schedule for scheduled tasks. Scheduled tasks are CSV files the users uploaded to be run at a given time. We already have the information of job size (number of notificaitons) and the scheduled time. The proactive scheduler scales up the number of instances before the jobs start.  
+
+The desired number of instances for the particular app is calculated as follow:
+
+```
+desired_instance_count=num_of_scheduled_jobs/request_per_instance/scheduled_job_scaling_factor
+and
+min_instance_count<=desired_instances_count<=max_instance_count
+```
+
+The `num_of_scheduled_jobs` is the number of notifications to be sent in that task. `Request_per_instance` gives a first indication on the number of jobs that can be handled by one app instance. We can use `scheduled_job_scaling_factor ` to control how aggressively we want the scaling to be. Currently, this factor is set to 2 in Notify. 
+
+## The Scaling function ##
+The function `scale_paas_apps` uses the `desired_instance_count` factor calculated to scale up or down the number of instances for the app. Below is the Psuedocode for the autoscaler. 
+
+```
+if desired_instance_count > current_instance_count
+	scale up to desired_instance_count
+else if time since last scale up > 5 mins
+	current_instance_count = current_instance_count - 1
+```
+
+The apps are instantly scaled up to the desired_instance_count if the current number of running instances is lower. However, on scaling down, a cool down period of 5 mins is allowed for the current apps to be finished before an instance is shut down. 
+
+## Minimum and maximum instance counts ##
+The `min_instance_count` and `max_instance_count` used to bound the `desired_instance_count` is defined differently in Notify's environment as below:
+
+| App | Queues / Load balancer | msg per instance | default in prod | Preview | Staging | Production |
+| --- | ------ | ---------------- | --------- | ------- | ------- | ----------|
+| notify-delivery-worker-sender | send-sms, send-email, send-tasks | 250 | 4 | 1-2 | 4-20 | 4-20 |
+| notify-delivery-worker | notify, retry, process-job, notify-internal-tasks, retry-tasks, job-tasks, periodic-tasks | 250 | 2 | 1-1 | 2-5 | 2-5 |
+| notify-delivery-worker-database | db-sms, db-email, db-letter, database-tasks | 250 | 2 | 1-2 | 2-20 | 2-20 |
+| notify-delivery-worker-periodic | periodic, statistics, periodic-tasks, statistics-tasks | 250 | 2 | 1-1 | 2-5 | 2-5 |
+| notify-delivery-worker-research | research-mode, research-mode-tasks | 250 | 2 | 1-1 | 2-5 | 2-5 |
+| notify-delivery-worker-priority | priority, priority-tasks | 250 | 2 | 1-1 | 2-5 | 2-5 |
+| notify-api | notify-paas-proxy | 1500 | 4 | 1-2 | 2-20 | 2-20 |
```

Comment:
Nice table!

Couple of notes: Probably better if we create two distinct queues and load balancer columns just to be explicit on which apps are scaled based on which. Same for `msg per instance` column to split in message per instance and requests per instance.

Although, if the resulting table is too wide we could keep the same number of columns and just add some kind of indication for the load balancer/request per instance row.

On the other hand, Staging and Production columns can be merged to one since we [made the same](https://github.com/alphagov/notifications-paas-autoscaler/pull/10/commits/b9d258c77455cd7552d48553b445d0396f5aec4e)